lib/spack/llnl/util/multiproc.py

```diff
@@ -8,25 +8,48 @@
 than multiprocessing.Pool.apply() can.  For example, apply() will fail
 to pickle functions if they're passed indirectly as parameters.
 """
-from multiprocessing import Process, Pipe, Semaphore, Value
+import functools
+from multiprocessing import Semaphore, Value
 
-__all__ = ['spawn', 'parmap', 'Barrier']
+__all__ = ['Barrier']
 
 
-def spawn(f):
-    def fun(pipe, x):
-        pipe.send(f(x))
-        pipe.close()
-    return fun
+def deferred(func):
+    """Package a function call into something that can be invoked
+    at a later moment.
 
+    Args:
+        func (callable): callable that must be deferred
 
-def parmap(f, elements):
-    pipe = [Pipe() for x in elements]
-    proc = [Process(target=spawn(f), args=(c, x))
-            for x, (p, c) in zip(elements, pipe)]
-    [p.start() for p in proc]
-    [p.join() for p in proc]
-    return [p.recv() for (p, c) in pipe]
+    Returns:
+        Deferred version of the same function
+    """
+    @functools.wraps(func)
+    def _impl(*args, **kwargs):
+        def _deferred_call():
+            return func(*args, **kwargs)
+        return _deferred_call
+    return _impl
+
+
+def invoke(f):
+    return f()
+
+
+def execute(command_list, map_fn=map, transformation=invoke):
```

Comment:
> Maybe I'm just used to map, but I think it's a decent abstraction... and a more familiar one. 

Agreed. With `execute` I tried to make the business logic more evident at call site by hiding the deferred call - couldn't do much though to hide the `mapping` function. Since the gain is any case small (and not without contention) I think it's right to remove it.

Reply:
> Why can't we assemble a matrix of arguments then pass it to functions rather than deferring all the calls? I like defer, but I think separating the data from the calls separates the concerns more, especially if we want to be portable between a thread pool and a process pool -- in the latter case you have to pickle arguments.

**TL;DR** I'll look into how hard is to do that. I mean, conceptually it's not that difficult, but if we need to use a `ProcessPool` I need to check which arguments are involved and whether they can be pickled or not. 

---

Right now we are basically farming out system calls. Before implementing the PR, I checked this small example on a 4 cores laptop:
```console
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux

>>> import multiprocessing.pool
>>> import subprocess 
>>> import timeit
>>>
>>> def foo():
...     p = subprocess.Popen(['sleep', '1'])
...     p.communicate()
...     return 'Done'
>>> tp = multiprocessing.pool.ThreadPool()
>>> l = [foo] * 8
>>> timeit.timeit('print(tp.map(lambda x: x(), a))', globals={'a': a, 'tp': tp}, number=2)
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
4.041111436999927
>>>
>>> timeit.timeit('print(list(map(lambda x: x(), a)))', globals={'a': a}, number=2)
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
16.08182063300046
```
to be confident that a `ThreadPool` could parallelize this part. Any line of Python code we add to `foo` I think it will be effectively serialized by a `ThreadPool` because of the GIL. If the overhead of that part is non negligible we'll need to move to a `ProcessPool` - and that's where the funny `pickle` part starts.

Reply:
Done in 5448ef0. I stayed with a `ThreadPool` for now, but tried a `Pool` with Python 3.6 and it works.